### PR TITLE
porting three subroutine to c++

### PR DESCRIPTION
**Pull request by xyuan**
Porting 
ice_cldliq_wet_growth
ice_nucleation
ice_relaxation_timescale
subroutine to c++

BFB

### CODE REVIEW

File: components/cam/src/physics/cam/micro_p3.F90 [hunk 1]
```diff
@@ -2333,18 +2337,30 @@ subroutine ice_cldliq_wet_growth(rho,t,pres,rhofaci,    &
...
+                                  log_wetgrowth,qrcol,qccol,qwgrth,nrshdr,qcshd)
+       return
+    endif
+#endif
 
    if (qitot_incld.ge.qsmall .and. qc_incld+qr_incld.ge.1.e-6_rtype .and. t.lt.zerodegc) then
       qsat0  = 0.622_rtype*e0/(pres-e0)
 
-      qwgrth = ((f1pr05 + f1pr14*sc**thrd*(rhofaci*rho/mu)**0.5_rtype)*       &
+      qwgrth = ((f1pr05 + f1pr14*bfb_cbrt(sc)*bfb_pow((rhofaci*rho/mu),0.5_rtype))*       &
```
Comment (ambrad) [inline components/cam/src/physics/cam/micro_p3.F90]: Here instead of bfb_pow(.,0.5) you should make a bfb_sqrt function, since we'll use std::sqrt in C++.

File: components/cam/src/physics/cam/micro_p3.F90 [hunk 2]
```diff
@@ -2388,10 +2408,18 @@ subroutine calc_ice_relaxation_timescale(rho,t,rhofaci,     &
...
+                                      f1pr05,f1pr14,dv,mu,sc,qitot_incld,nitot_incld, &
+                                      epsi,epsi_tot)
+       return
+    endif
+#endif
 
 
    if (qitot_incld.ge.qsmall .and. t.lt.zerodegc) then
-      epsi = ((f1pr05+f1pr14*sc**thrd*(rhofaci*rho/mu)**0.5_rtype)*2._rtype*pi* &
+      epsi = ((f1pr05+f1pr14*bfb_cbrt(sc)*(rhofaci*rho/mu)**0.5_rtype)*2._rtype*pi* &
```
Comment (ambrad) [inline components/cam/src/physics/cam/micro_p3.F90]: Here you'll need to use bfb_sqrt for the `**0.5` part.

File: components/cam/src/physics/cam/micro_p3.F90 [hunk 3]
```diff
@@ -2643,10 +2673,20 @@ subroutine ice_nucleation(t,inv_rho,nitot,naai,supi,odt,log_predictNc,    &
...
+    endif
+#endif
+
    if ( t .lt.icenuct .and. supi.ge.0.05_rtype) then
       if(.not. log_predictNc) then
-!         ! dum = exp(-0.639+0.1296*100.*supi(i,k))*1000.*inv_rho(i,k)  !Meyers et al. (1992)
-         dum = 0.005_rtype*exp(0.304_rtype*(zerodegc-t))*1000._rtype*inv_rho   !Cooper (1986)
+         ! dum = exp(-0.639+0.1296*100.*supi(i,k))*1000.*inv_rho(i,k)  !Meyers et al. (1992)
+         dum = 0.005_rtype*bfb_exp(0.304_rtype*(zerodegc-t))*1000._rtype*inv_rho   !Cooper (1986)
+
```
Comment (ambrad) [inline components/cam/src/physics/cam/micro_p3.F90]: I reocmmend you make a separate PR with just the non-BFB F90 changes. This very small PR will be non-BFB. Then make a second PR that has the rest of the changes. This second PR will be BFB w.r.t. master, since master will have the F90 changes from the first PR.

File: components/scream/src/physics/p3/p3_functions_ice_cldliq_wet_growth_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+   const Spack e0 = polysvp1(zerodeg, 0.0);
+   Spack qsat0{0.0};
+   Spack dum{0.0};
+   Spack dum1{0.0};
+
+   qsat0.set(any_if,
+             sp(0.622)*e0/(pres-e0));
+  
+   qwgrth.set(any_if,
+             ((f1pr05+f1pr14*pack::cbrt(sc)*pack::pow(rhofaci*rho/mu, half))*
```
Comment (ambrad) [inline components/scream/src/physics/p3/p3_functions_ice_cldliq_wet_growth_impl.hpp]: Here you should use sqrt instead of pow.

File: components/scream/src/physics/p3/p3_functions_ice_cldliq_wet_growth_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+   dum.set(any_if,
+           pack::max(zero, (qccol+qrcol)-qwgrth));
+
+   auto const dum_ge_small = dum >= qsmall;
+
+   nrshdr.set(any_if && dum_ge_small,
+              nrshdr+dum*sp(1.923e+6));
+
+   dum1.set(any_if_col && dum_ge_small,
```
Comment (ambrad) [inline components/scream/src/physics/p3/p3_functions_ice_cldliq_wet_growth_impl.hpp]: Factor out common masks into a mask variable.

File: components/cam/src/physics/cam/micro_p3.F90 (hunk 3 above)
Comment (xyuan) [reply components/cam/src/physics/cam/micro_p3.F90]: OK, I will do it tomorrow, thanks.

**Comment by xyuan** [comment]
close this PR, and replace it with 3 PRs
